src/dbnode/storage/index.go

```diff
@@ -716,77 +723,196 @@ func (i *nsIndex) Query(
 	opts index.QueryOptions,
 ) (index.QueryResults, error) {
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		return index.QueryResults{}, errDbIndexUnableToQueryClosed
 	}
 
-	// override query response limit if needed.
+	// Track this as an inflight query that needs to finish
+	// when the index is closed
+	i.queriesWg.Add(1)
+	defer i.queriesWg.Done()
+
+	// Enact overrides for query options
+	opts = i.overriddenOptsForQueryWithLock(opts)
+	timeout := i.timeoutForQueryWithLock(ctx)
+
+	// Retrieve blocks to query, then we can release lock
+	// NB(r): Important not to block ticking, and other tasks by
+	// holding the RLock during a query
+	blocks, err := i.blocksForQueryWithLock(xtime.NewRanges(xtime.Range{
+		Start: opts.StartInclusive,
+		End:   opts.EndExclusive,
+	}))
+
+	// Can now release the lock and execute the query without holding the lock
+	i.state.RUnlock()
+
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	var (
+		start      = i.nowFn()
+		deadline   = start.Add(timeout)
+		exhaustive = true
+		results    = i.newConcurrentResults(ctx)
+		multiErr   = xerrors.NewMultiError()
+		workers    = i.opts.QueryIDsWorkerPool()
+		wg         sync.WaitGroup
+	)
+	for _, block := range blocks {
+		// Capture block for async query execution below
+		block := block
+
+		// terminate early if we know we don't need any more results
+		if opts.Limit > 0 && results.Size() >= opts.Limit {
+			results.Lock()
+			exhaustive = false
+			results.Unlock()
+			break
+		}
+
+		execBlockQuery := func() {
+			currExhaustive, err := block.Query(query, opts, results)
+			if err != index.ErrUnableToQueryBlockClosed {
+				// NB(r): Because we query this block outside of the lock, it's possible this
+				// block may get closed if it slides out of retention, in that case
+				// those results are no longer considered valid and outside of retention
+				// regardless, so this is a non-issue
+				err = nil
+			}
+
+			results.Lock()
+			defer results.Unlock()
+
+			if err != nil {
+				multiErr = multiErr.Add(err)
+				return
+			}
+
+			if !currExhaustive {
+				// i.e. block had more data but we stopped early, we know
+				// we have hit the limit and don't need to query any more.
+				exhaustive = false
+				return
+			}
+		}
+
+		if timeoutApplied := timeout > 0; !timeoutApplied {
+			// No timeout, just wait blockingly for a worker
+			wg.Add(1)
+			workers.Go(func() {
```

Comment:
Can we add a comment here (or somewhere else) with a TODO to replace this with the pooled worker pool? Previously we were creating goroutines proportional to number of queries, now its proportional to number of queries * number of blocks. Still a probably small number, but I can see it having some impact if these stacks need to grow to 4/8KiB

Reply:
In the comment you can basically say: "Swap this out with PooledWorkerPool once we have an implementation with timeout support" or something like that